app/controllers/application_controller.rb

```diff
@@ -61,15 +61,28 @@ def valid_help_actions
 
   def confirm_two_factor_setup
     user = request.env["warden"].user
-    return unless user &&
-      user.need_two_factor_authentication?(request) &&
-      !user.totp_enabled?
 
     if Rails.configuration.enable_enhanced_2fa_experience
+      return unless user
+
+      unless user.second_factor_method.nil?
+        # Handled by Devise's controller.
+        return if user.second_factor_method == "app"
+
+        # Handled by us.
+        redirect_to users_two_factor_authentication_email_path
+        return
+      end
+
       # TODO: Replace with route helper once fully migrated to enhanced 2FA.
       redirect_to "/users/two_factor_authentication/setup"
-    else
-      redirect_to users_two_factor_authentication_setup_path
+      return
     end
+
+    return unless user &&
+      user.need_two_factor_authentication?(request) &&
+      !user.totp_enabled?
```

Comment:
all this logic should live in the user model (`user.rb`) in a method that indicates what this is checking for (user exists, and needs 2FA but doesn't have app 2FA?)

Reply:
@koetsier was looking at refactoring this bit anyway, so happy to take your comment onboard again on that PR.